sql: TestIndexBackfillFractionTracking failed [no valid leaseholder for a range]

sql.TestIndexBackfillFractionTracking [failed](https://tanzanite.cluster.engflow.com/invocations/default/b5b2fc92-0625-498a-9718-5b1f6b13d385?testReportRun=3&testReportShard=15&testReportAttempt=1#targets-Ly9wa2cvc3FsOnNxbF90ZXN0) on master @ [0b6de9c809f8a4df2ba943a8c9dd023adb03b01d](https://github.com/cockroachdb/cockroach/commits/0b6de9c809f8a4df2ba943a8c9dd023adb03b01d):

Fatal error:

```
panic: test timed out after 1h16m37s
	github.com/cockroachdb/cockroach/pkg/util/log/file_log_gc.go:25 +0xd3
running tests:
created by github.com/cockroachdb/cockroach/pkg/util/log.ApplyConfig in goroutine 23074
	github.com/cockroachdb/cockroach/pkg/util/log/flags.go:319 +0x1d65

```
Stack: 

```
goroutine 23674 [select, 5 minutes]:
	TestIndexBackfillFractionTracking (1h15m54s)
github.com/cockroachdb/cockroach/pkg/sql/sqlstats/persistedsqlstats.(*PersistedSQLStats).startSQLStatsFlushLoop.func1({0xdfb9248, 0xc00a675d10})
```
<details><summary>Log preceding fatal error</summary>
<p>

```
	github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:189 +0x2e5
github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*dbAdapter).RangeFeed(0xc003283d28, {0xdfb9210, 0xc00332f950}, {0xc0018c86f0, 0x1, 0x1}, {0xc004e31fd0?, 0x2097c6d?}, 0xc008755e60, {0xc002cff600, ...})
	github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed/db_adapter.go:79 +0xed
github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).run.func1({0xdfb9210, 0xc00332f950})
	github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed/rangefeed.go:313 +0x186
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait.Group.GoCtx.func1()
	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:168 +0x47
golang.org/x/sync/errgroup.(*Group).Go.func1()
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:78 +0x92
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 23320
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:75 +0x125

goroutine 23193 [select]:
github.com/cockroachdb/cockroach/pkg/gossip.(*Gossip).manage.func1({0xdfb9248, 0xc005d774a0})
	github.com/cockroachdb/cockroach/pkg/gossip/pkg/gossip/gossip.go:1239 +0x45b
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x263
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 23074
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x69d

goroutine 23165 [select]:
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).startGossip.func3({0xdfb9248, 0xc005d51650})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_gossip.go:146 +0x3ac
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x263
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 23194
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x69d

goroutine 24741 [semacquire, 75 minutes]:
sync.runtime_Semacquire(0xc002bfc490?)
	GOROOT/src/runtime/sema.go:62 +0x25
sync.(*WaitGroup).Wait(0xc002bfc488)
	GOROOT/src/sync/waitgroup.go:116 +0xa5
golang.org/x/sync/errgroup.(*Group).Wait(0xc002bfc480)
	golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:56 +0x38
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.Wait({0xc002bfc480?, {0xdfb9210?, 0xc0054d1090?}})
	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:144 +0x65
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait({0xdfb9248, 0xc006582d80}, {0xc006fcbb98, 0x2, 0x97ed9a0?})
	github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:192 +0x199
github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).run(0xc0047f3ce0, {0xdfb9248, 0xc006582d80}, {0xe00c660, 0xc00640af00})
	github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed/rangefeed.go:319 +0xc4e
github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).Start.func2({0xdfb9248, 0xc00640b1a0})
	github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed/rangefeed.go:230 +0x565
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484 +0x263
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 23074
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:475 +0x69d

goroutine 23075 [select, 75 minutes]:
github.com/cockroachdb/cockroach/pkg/util/log.(*fileSink).gcDaemon(0xc00344efc0, {0xdfb9210, 0xc004f902d0})
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=15</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestIndexBackfillFractionTracking.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-36789

--- a/pkg/sql/mvcc_backfiller_test.go
+++ b/pkg/sql/mvcc_backfiller_test.go
@@ -671,10 +671,11 @@ func splitIndex(
 			return err
 		}
 
-		holder, err := tc.FindRangeLeaseHolder(rangeDesc, nil)
+		li, _, err := tc.FindRangeLeaseEx(ctx, rangeDesc, nil)
 		if err != nil {
 			return err
 		}
+		holder := li.CurrentOrProspective().Replica
 
 		_, rightRange, err := tc.Server(int(holder.NodeID) - 1).SplitRange(pik)
 		if err != nil {

--- a/pkg/testutils/serverutils/test_cluster_shim.go
+++ b/pkg/testutils/serverutils/test_cluster_shim.go
@@ -153,6 +153,21 @@ type TestClusterInterface interface {
 		hint *roachpb.ReplicationTarget,
 	) (roachpb.ReplicationTarget, error)
 
+	// FindRangeLeaseEx returns information about a range's lease. As opposed to
+	// FindRangeLeaseHolder, it doesn't check the validity of the lease; instead it
+	// returns a timestamp from a node's clock.
+	//
+	// If hint is not nil, the respective node will be queried. If that node doesn't
+	// have a replica able to serve a LeaseInfoRequest, an error will be returned.
+	// If hint is nil, the first node is queried. In either case, if the returned
+	// lease is not valid, it's possible that the returned lease information is
+	// stale - i.e. there might be a newer lease unbeknownst to the queried node.
+	FindRangeLeaseEx(
+		ctx context.Context,
+		rangeDesc roachpb.RangeDescriptor,
+		hint *roachpb.ReplicationTarget,
+	) (_ roachpb.LeaseInfo, now hlc.ClockTimestamp, _ error)
+
 	// TransferRangeLease transfers the lease for a range from whoever has it to
 	// a particular store. That store must already have a replica of the range. If
 	// that replica already has the (active) lease, this method is a no-op.

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -1277,15 +1277,7 @@ func (tc *TestCluster) FindRangeLease(
 	return l.CurrentOrProspective(), now, err
 }
 
-// FindRangeLeaseEx returns information about a range's lease. As opposed to
-// FindRangeLeaseHolder, it doesn't check the validity of the lease; instead it
-// returns a timestamp from a node's clock.
-//
-// If hint is not nil, the respective node will be queried. If that node doesn't
-// have a replica able to serve a LeaseInfoRequest, an error will be returned.
-// If hint is nil, the first node is queried. In either case, if the returned
-// lease is not valid, it's possible that the returned lease information is
-// stale - i.e. there might be a newer lease unbeknownst to the queried node.
+// FindRangeLeaseEx is part of TestClusterInterface.
 func (tc *TestCluster) FindRangeLeaseEx(
 	ctx context.Context, rangeDesc roachpb.RangeDescriptor, hint *roachpb.ReplicationTarget,
 ) (_ roachpb.LeaseInfo, now hlc.ClockTimestamp, _ error) {

